pkg/controllers/state/cluster.go

```diff
@@ -183,6 +187,40 @@ func (c *Cluster) IsNodeNominated(providerID string) bool {
 	return false
 }
 
+// IsNodeUnderutilized returns true if the last pod event on the node was at least consolidateAfter
+// in the past. Also returns a reason to inform the user on how this was calculated.
+func (c *Cluster) IsNodeUnderutilized(providerID string, consolidateAfter time.Duration) (bool, time.Duration, string) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	reason := ""
+	lastPodEventTime, ok := c.lastPodEvents[providerID]
+	if !ok {
+		// If the node doesn't have a pod event, we assume it's because either:
+		// 1. Karpenter has restarted
+		// 2. No pods have ever scheduled to the node
+		//
+		// So treat the last pod time as the more recent of
+		// 1. When Karpenter restarted
+		// 2. When the Node was created
+		if c.clock.Since(c.initTime) < c.clock.Since(c.nodes[providerID].NodeClaim.CreationTimestamp.Time) {
+			lastPodEventTime = c.initTime
+			reason = "karpenter controller initialization time"
+		} else {
+			lastPodEventTime = c.nodes[providerID].NodeClaim.CreationTimestamp.Time
+			reason = "node creation time"
+		}
+	}
+	timeSince := c.clock.Since(lastPodEventTime)
+	// If the lastPodEventTime is somehow in the future, treat the duration as 0
+	if timeSince < 0 {
+		timeSince = 0 * time.Second
+	}
+	return timeSince >= consolidateAfter,
```

Comment:
IMO: we could collapse a lot of this down into just getting the last time that there was movement on the node from any pod